src/instrumenter/annotations.ts

```diff
@@ -546,6 +608,163 @@ export function gatherContractAnnotations(
+                const target = [...targets][0];
+
+                if (target instanceof FunctionDefinition && args.length > 0) {
+                    const params = target.vParameters.vParameters;
+
+                    mapArgsToParams(signature, args, params, localAliases);
+                }
+
+                for (const { expression, message } of properties) {
+                    const annotation = parseAnnotation(expression, target, compilerVersion);
+
+                    assert(
+                        annotation instanceof SProperty,
+                        `Only properties are allowed to be defined in macros. Unsupported annotation: ${expression}`
+                    );
+
+                    /**
+                     * Update variable names with supplied values from macro instruction.
+                     */
+                    annotation.expression.walk(walker);
+                    annotation.label = message;
+                    annotation.prefix = "#";
+
+                    /**
+                     * @todo This is temporary hack. Need to find a better way to do this.
+                     */
+                    const dummyDoc = new StructuredDocumentation(
+                        0,
+                        "0:0:0",
+                        "StructuredDocumentation",
+                        annotation.prefix + annotation.pp()
+                    );
+
+                    const metaToInject = new PropertyMetaData(
+                        dummyDoc,
+                        target,
+                        dummyDoc.text,
+                        annotation,
+                        0,
+                        dummyDoc.text
+                    );
+
+                    validateAnnotation(target, metaToInject);
+
+                    const metasToInject = injections.get(target);
+
+                    if (metasToInject === undefined) {
+                        injections.set(target, [metaToInject]);
+                    } else {
+                        metasToInject.push(metaToInject);
+                    }
+                }
+            }
+
+            /**
+             * Remove macro annotation as it will not pass type-checking.
+             * It is now represented by other annotations in `injections`,
+             * so there is no further use for macro annotation itself.
+             */
+            metas.splice(m, 1);
+        }
+    }
+
+    /**
+     * Merge injected annotations with original annotations.
+     */
+    for (const [target, metasToInject] of injections) {
+        const metas = annotationMap.get(target);
+
+        if (metas === undefined) {
+            annotationMap.set(target, metasToInject);
+        } else {
+            metas.push(...metasToInject);
+        }
+    }
+
+    /**
+     * @todo list:
+     * [TODO]   1. Source locations are needed for property maps.
+     * [TODO]   2. Figure out how to use macro variable types.
```

Comment:
What do you mean by variable types?

Reply:
Macro variables can be defined with the types, however those types are never user right now. There is no typechecking that macro call values are matching macro variable types.

Reply:
So I am thinking lets go the route of C macros - basically we don't check the types of variables at macro instantiation, and instead we just type check the inlined invariants. This may result in weird type errors for now.

However if we fix the node sources to keep track which macro definition an annotation came from, then when we encounter type errors we will correctly report that the type error was in the body of annotation X instantiated in location Y as part of macro Z. This should be enough information for now for users to make progress.

Reply:
Fixed with 072f378